README.md
# hijackresponse

[![npm version](https://badge.fury.io/js/hijackresponse.svg)](https://www.npmjs.com/package/hijackresponse)
[![Build Status](https://travis-ci.org/gustavnikolaj/hijackresponse.svg)](https://travis-ci.org/gustavnikolaj/hijackresponse)
[![Coverage Status](https://coveralls.io/repos/gustavnikolaj/hijackresponse/badge.svg?branch=master&service=github)](https://coveralls.io/github/gustavnikolaj/hijackresponse?branch=master)
[![Dependency Status](https://david-dm.org/gustavnikolaj/hijackresponse.svg)](https://david-dm.org/gustavnikolaj/hijackresponse)

Module that allows you to rewrite HTTP responses from middleware further down
the stack, such as static providers, HTTP proxies etc.

Requires node v8 or later.

This module is the spiritual successor to
[express-hijackresponse](https://github.com/papandreou/express-hijackresponse)
attempting to solve the same goals. The differences from the original module are
primarily that the API is slightly different, there's no direct coupling to
express and it supports streams2.

It's mostly useful for content filters. The original use case is injecting an
inline JavaScript into all HTML responses in
[LiveStyle](https://github.com/One-com/livestyle). It is also used in a series
of transpiler and preprocessing middleware:

- [express-compiless](https://github.com/papandreou/express-compiless)
- [express-processimage](https://github.com/papandreou/express-processimage)
- [express-extractheaders](https://github.com/papandreou/express-extractheaders)
- [express-autoprefixer](https://github.com/gustavnikolaj/express-autoprefixer)

## Installation

```
$ npm install hijackresponse
```

## Usage

```js
var express = require("express");
var hijackResponse = require("hijackresponse");

var app = express();

app.use((req, res, next) => {
  hijackResponse(res, next).then(({ readable, writable }) => {
    // Don't hijack HTML responses:
    if (/^text\/html/.test(res.getHeader("Content-Type"))) {
      return readable.pipe(writable);
    }

    res.setHeader("X-Hijacked", "yes!");
    res.removeHeader("Content-Length");

    readable.pipe(transformStream).pipe(writable);
  });
});
```

## API

### Function `hijackResponse()`

```
hijackResponse(res[, cb]) => Promise<HijackedReponse>
```

The `hijackResponse` function takes one required argument - the response object
which is the target of the hijacking. The second optional argument, is a
callback to be called when the hijacking preparations are done; this will mostly
be used when you are working with express. You can also decide to call the
callback afterwards if you prefer. The following two examples are equivalent:

```js
app.use((req, res, next) => {
  hijackResponse(res, next).then(() => { /* ... */});
});

app.use((req, res, next) => {
  hijackResponse(res).then(() => { /* ... */});
  next();
});
```

The first example is easier to work with when you are working with async/await:

```js
// Using express-promise-router or equivalent

app.use(async (req, res, next) => {
  const hijackedResponse = await hijackResponse(res, next);

  // ... do something with the hijacked reponse.
})
```

### Object `hijackedResponse`

```
{
  readable: NodeJS ReadableStream,
  writable: NodeJS Writable,
  destroyAndRestore: Function
}
```

The resolution value of the Promise returned from calling `hijackResponse`.

- `readable` is a readable stream containing the captured response body.
- `writable` is a writable stream which will be sent to the client.
- `destroyAndRestore` is a function that destroys the readable stream, and
  restores the original res.

Everything written to `res` in other handlers are captured, so if you want to
delegate to the express errorhandler you need to call `destroyAndRestore` before
doing so. Calling `destroyAndRestore` will undo the hijack, and destroy the `readable`
stream, meaning that all data written to it so far is discarded.

```js
app.use((req, res, next) => {
  hijackResponse(res, next).then((hijackedResponse) => {
    hijackedResponse.destroyAndRestore();
    return next(new Error('Something bad happened'));
  });
});
```

If you don't call `destroyAndRestore` before passing the error to next, the
errorhandlers output will become available on the `readable`-stream instead of
being sent to the client as intended.

## License

This module is published under the ISC License. See the `LICENCE` file for
additional details.


CHANGELOG.md
### v5.0.0 (2020-09-28)

#### Pull requests

- [#29](https://github.com/gustavnikolaj/hijackresponse/pull/29) Use an actual Proxy to create proxyRes ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [#30](https://github.com/gustavnikolaj/hijackresponse/pull/30) Proof of concept of promise API ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [#24](https://github.com/gustavnikolaj/hijackresponse/pull/24) Split stream rewrite ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [#21](https://github.com/gustavnikolaj/hijackresponse/pull/21) Supporting node.js v12 ([Gustav Nikolaj Olsen](mailto:gno@one.com))

#### Commits to master

- [Update example and add API docs](https://github.com/gustavnikolaj/hijackresponse/commit/942a00577a4f71a1689f6037c020fed0fd620c9d) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Setup automatic changelog generation](https://github.com/gustavnikolaj/hijackresponse/commit/94a197ff519d933ae5fa01916e33785d93ab98cf) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Use a Symbol for marking hijacked responses](https://github.com/gustavnikolaj/hijackresponse/commit/d41a83bb570fff9a66236f98e23f6c8a5650fb15) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [5.0.0-rc.2](https://github.com/gustavnikolaj/hijackresponse/commit/2f4f1c308774d84c7b2c3c14ca0f35cd096053f1) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Add jsdoc comments to hijackResponse function](https://github.com/gustavnikolaj/hijackresponse/commit/55bad0747c7b4522fd901d9c1a7dc991793a23e3) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [+31 more](https://github.com/gustavnikolaj/hijackresponse/compare/v4.0.1...v5.0.0)

### v4.0.1 (2020-09-14)

#### Pull requests

- [#18](https://github.com/gustavnikolaj/hijackresponse/pull/18) Whitelist package contents ([Matthias Kunnen](mailto:matthias.kunnen@gmail.com))

#### Commits to master

- [Update code formatting](https://github.com/gustavnikolaj/hijackresponse/commit/91f7f6ac82312632a8f58f528f28d040eb9d0082) ([Gustav Nikolaj Olsen](mailto:gno@one.com))

### v4.0.0 (2019-03-04)

- [Remove warning on being untested on node 8 and above](https://github.com/gustavnikolaj/hijackresponse/commit/55bceb85cf27f2f9d9fbb9ea38325276ef6e6303) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Test on node 6, 8 and 10](https://github.com/gustavnikolaj/hijackresponse/commit/5a7f6391389770a04669756643aeb03e12aa3d87) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Make sure emit gets copied over](https://github.com/gustavnikolaj/hijackresponse/commit/1d587ca2519ad8fbd29527dc16689811a09691b0) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Trim editorconfig](https://github.com/gustavnikolaj/hijackresponse/commit/3d5b3736b25917bf0f49c921058f615f80baecdd) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [Replace istanbul with nyc](https://github.com/gustavnikolaj/hijackresponse/commit/8a59f4bd5a481b91823c504bc7ca9cf18bf57382) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [+13 more](https://github.com/gustavnikolaj/hijackresponse/compare/v3.0.0...v4.0.0)

### v3.0.0 (2017-04-25)

- [fix handling of status messages in res.writeHead \(\#13\) \(fixes \#12\)](https://github.com/gustavnikolaj/hijackresponse/commit/55631c7993baef698d8797b4c286f6879ca7d744) ([Gustav Nikolaj](mailto:gustavnikolaj@gmail.com))

### v2.0.1 (2016-12-02)

#### Pull requests

- [#11](https://github.com/gustavnikolaj/hijackresponse/pull/11) Preserving "close" Event Listeners On Response ([Pablo Sichert](mailto:mail@pablosichert.de))

#### Commits to master

- [Travis: Drop node.js 0.10 support, add 6 and 7 instead.](https://github.com/gustavnikolaj/hijackresponse/commit/a6fba84754b32bdb5a69a370a9aad1ce02ae27b5) ([Andreas Lind](mailto:andreas@one.com))

### v2.0.0 (2016-08-05)

- [remove backpressure support \(ref \#10\)](https://github.com/gustavnikolaj/hijackresponse/commit/000717c70543b362be85368c38522275c2032d5e) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [accept that the unchunked body is either an empty buffer or undefined](https://github.com/gustavnikolaj/hijackresponse/commit/4bb1eb842e2358a27c8ca7fa9360917d0fd7dc66) ([Gustav Nikolaj Olsen](mailto:gno@one.com))
- [fix lint errors](https://github.com/gustavnikolaj/hijackresponse/commit/cb8b349792d532ac66892d46a2c0ee3d8cf4f939) ([Gustav Nikolaj Olsen](mailto:gno@one.com))

### v1.1.0 (2016-04-15)

#### Pull requests

- [#7](https://github.com/gustavnikolaj/hijackresponse/pull/7) set right string name ([Holger Schauf](mailto:holger.schauf@googlemail.com))
- [#6](https://github.com/gustavnikolaj/hijackresponse/pull/6) Wrong Regex ([Holger Schauf](mailto:holger.schauf@googlemail.com))

#### Commits to master

- [Consolidate with fix\/nobackpressure by introducing a disableBackpressure option.](https://github.com/gustavnikolaj/hijackresponse/commit/1a862faee84c6bd8c304ab5e71553c4a7088652e) ([Andreas Lind](mailto:andreas@one.com))
- [Added failing test \[ci skip\].](https://github.com/gustavnikolaj/hijackresponse/commit/e703ef5377134446f1e2c05f310a34926428611b) ([Andreas Lind](mailto:andreas@one.com))
- [Implement .destroyHijacked method.](https://github.com/gustavnikolaj/hijackresponse/commit/a4e3e6788f821836238f24ae39f39f07e03fdc01) ([Andreas Lind](mailto:andreas@one.com))
- [Update unexpected et al., remove peerDeped modules that are also deps.](https://github.com/gustavnikolaj/hijackresponse/commit/1d92e59b7cb2b69650ffaf78398b27b3a90f1e62) ([Andreas Lind](mailto:andreas@one.com))

### v1.0.4 (2016-03-30)

- [#5](https://github.com/gustavnikolaj/hijackresponse/pull/5) Relay the 'close' event from the original res to the hijacked response. ([Andreas Lind](mailto:andreas@one.com))

### v1.0.3 (2016-02-18)

#### Pull requests

- [#4](https://github.com/gustavnikolaj/hijackresponse/pull/4) Fix repository link in package.json. ([Hugo Josefson](mailto:hugo@josefson.org))

#### Commits to master

- [add homepage and bugs props to package.json](https://github.com/gustavnikolaj/hijackresponse/commit/b4eb7f55922eb8c30c06b40c656eca49014a293f) ([Gustav Nikolaj Olsen](mailto:gno@one.com))

test/poc.js
var Writable = require('stream').Writable
var Readable = require('stream').Readable
var Transform = require('stream').Transform

var Stream = require('stream').Stream

var convertToUpperCaseTransformStream = new Transform()
convertToUpperCaseTransformStream._transform = function (chunk, encoding, cb) {
  if (encoding === 'buffer') {
    var result = []
    for (var i = 0; i < chunk.length; ++i) {
      var charCode = chunk[i]
      if (charCode > 96 && charCode < 123) {
        charCode = charCode - 32
      }
      result.push(charCode)
    }
    this.push(new Buffer(result))
    return cb()
  }
  return cb(new Error('chunk not a buffer!'))
}

//process.stdin
//  .pipe(convertToUpperCaseTransformStream)
//  .pipe(process.stdout)

function hijack (res, cb) {
  var origWrite = res.write
  var origEnd = res.end
  var origRes = res

  var hijackedRes = new Readable()
  hijackedRes.__proto__ = origRes

  var readableMethods = Object.keys(Readable.prototype)
  readableMethods.forEach(function (method) {
    hijackedRes[method] = Readable.prototype[method].bind(hijackedRes)
  })

  var _readReady = false
  var writeBuffer = []

  hijackedRes._read = function (size) {
    _readReady = true
  }

  res.write = function (chunk, encoding) {
    if (_readReady && writeBuffer.length > 0) {
      var newWriteBuffer = []
      writeBuffer.forEach(function (obj) {
        if (_readReady) {
          _readReady = hijackedRes.push(obj.chunk)
        } else {
          newWriteBuffer.push(obj)
        }
      })
      writeBuffer = newWriteBuffer
    }
    if (_readReady) {
      _readReady = hijackedRes.push(chunk)
    } else {
      writeBuffer.push({
        chunk: chunk,
        encoding: encoding
      })
    }
  }

  res.end = function (chunk, encoding) {
    if (chunk) {
      res.write(chunk, encoding)
    }
    res.write(null)
  }

  hijackedRes.write = function (chunk, encoding) {
    origWrite.call(origRes, chunk, encoding)
  }

  hijackedRes.end = function (chunk, encoding) {
    if (chunk) {
      origWrite.call(origRes, chunk, encoding)
    }
    origEnd.call(origRes)
  }

  return cb(null, hijackedRes)
}

var res = new Writable()

res._write = function (chunk, encoding, next) {
  if (encoding === 'buffer') {
    process.stdout.write(chunk)
    return next()
  }
  throw new Error('not a buffer!')
}

// hijack(res, function (err, res) {
//   res
//     .pipe(convertToUpperCaseTransformStream)
//     .pipe(res)
// })

hijack(res, function (err, res) {
  res.on('data', function (chunk) {
    console.log('on data', chunk.toString().replace('\n', '\\n'))
    setTimeout(function () { res.write(chunk) }, 2)
  }).on('end', function () {
    console.log('on end')
    setTimeout(function () { res.end() }, 4)
  })
})

res.write('foo')
res.write('bar')
res.end('\n')


test/express.js
/* global describe, it, before, after */
var http = require('http')
var expect = require('unexpected')
  .clone()
  .use(require('unexpected-express'))
  .use(require('unexpected-sinon'))
  .addAssertion('to yield response', function (expect, subject, value) {
    return expect(subject, 'to yield exchange', {
      request: 'GET /',
      response: value
    })
  })
var sinon = require('sinon')
var express = require('express')
var passError = require('passerror')
var hijackResponse = require('../')

describe('Express Integration Tests', function () {
  it('simple case', function () {
    var app = express()
      .use(function (req, res, next) {
        hijackResponse(res, passError(next, function (res) {
          var chunks = []
          res.on('data', function (chunk) {
            chunks.push(chunk)
          })
          res.on('end', function () {
            var result = Buffer.concat(chunks).toString('utf-8').toUpperCase()
            res.write(new Buffer(result))
            res.end()
          })
        }))
        next()
      })
      .use(function (req, res, next) {
        res.setHeader('Content-Type', 'text/plain')
        return res.end('foobar')
      })

    return expect(app, 'to yield response', {
      statusCode: 200,
      body: 'FOOBAR'
    })
  })
  describe('adapted from express-hijackresponse', function () {
    it('Create a test server that pipes the hijacked response into itself, then do a request against it (simple variant)', function () {
      var app = express()
        .use(function (req, res, next) {
          hijackResponse(res, passError(next, function (res) {
            res.pipe(res)
          }))
          next()
        })
        .use(function (req, res, next) {
          res.send('foo')
        })

      return expect(app, 'to yield response', 'foo')
    })
    it('Create a test server that pipes the hijacked response into itself, then do a request against it (streaming variant)', function () {
      var app = express()
        .use(function (req, res, next) {
          hijackResponse(res, passError(next, function (res) {
            res.pipe(res)
          }))
          next()
        })
        .use(function (req, res, next) {
          var num = 0
          res.setHeader('Content-Type', 'text/plain')
          function proceed () {
            if (num < 5) {
              res.write('foo')
              num += 1
              process.nextTick(proceed)
            } else {
              res.end('bar')
            }
          }
          proceed()
        })

      return expect(app, 'to yield response', 'foofoofoofoofoobar')
    })
    it('Create a test server that pipes the original response through a buffered stream, then do a request against it (simple variant)', function () {
      var app = express()
        .use(function (req, res, next) {
          hijackResponse(res, passError(next, function (res) {
            var bufferedStream = new (require('bufferedstream'))()
            res.pipe(bufferedStream)
            bufferedStream.pipe(res)
          }))
          next()
        })
        .use(function (req, res, next) {
          res.send('foo')
        })

      return expect(app, 'to yield response', 'foo')
    })
    it('Create a test server that pipes the original response through a buffered stream, then do a request against it (streaming variant)', function () {
      var app = express()
        .use(function (req, res, next) {
          hijackResponse(res, passError(next, function (res) {
            var bufferedStream = new (require('bufferedstream'))()
            res.pipe(bufferedStream)
            bufferedStream.pipe(res)
          }))
          next()
        })
        .use(function (req, res, next) {
          res.contentType('text/plain')
          res.write('foo')
          res.end('bar')
        })

      return expect(app, 'to yield response', 'foobar')
    })
    it('Create a test server that hijacks the response and passes an error to next(), then run a request against it', function () {
      var app = express()
        .use(function (req, res, next) {
          hijackResponse(res, passError(function (res) {
            res.unhijack(function (res) {
              next(new Error('Error!'))
            })
          }))
          next()
        })
        .use(function (req, res, next) {
          res.send('foo')
        })
        .use(require('errorhandler')({ log: false }))

      return expect(app, 'to yield response', 500)
    })
    it('Create a test server that hijacks the response and immediately unhijacks it, then run a request against it', function () {
      var app = express()
        .use(function (req, res, next) {
          hijackResponse(res, passError(next, function (res) {
            res.unhijack(true)
          }))
          next()
        })
        .use(function (req, res, next) {
          res.send('foo')
        })

      return expect(app, 'to yield response', 'foo')
    })
  })

  it('should work when hijacking a big response body and the compression middleware is present above the hijacking middleware', function () {
    return expect(
      express()
        .use(require('compression')())
        .use(function (req, res, next) {
          hijackResponse(res, passError(next, function (res) {
            res.pipe(res)
          }))
          next()
        })
        .use(express.static(__dirname)),
      'to yield exchange', {
        request: 'GET /bigfile.txt',
        response: { body: /^0{1999998}$/ }
      }
    )
  })
  describe('against a real server', function () {
    before(function () {
      var self = this
      var spy = sinon.spy()

      this.spy = spy

      return new Promise(function (resolve) {
        var app = express()
          .use(function (req, res, next) {
            res.on('close', function () {
              spy()
              self.resolve()
            })

            next()
          })
          .use(function (req, res, next) {
            hijackResponse(res, passError(next, function (res) {
              res.pipe(res)
            }))

            next()
          })
          .use(function (req, res) {
            res.write('foo')
          })

        var server = http.Server(app)
        server.listen(0, function () {
          resolve(server)
        })
      }).then(function (server) {
        self.server = server
      })
    })
    it('should not prevent "close" events registered on res from firing when hijacking', function () {
      var self = this
      var port = this.server.address().port
      var spy = this.spy

      return new Promise(function (resolve) {
        self.resolve = resolve

        var options = {
          port: port
        }

        var request = http.request(options, function (res) {
          request.abort()
        })

        request.end()
      }).then(function () {
        expect(spy, 'was called')
      })
    })
    after(function () {
      this.server.close()
    })
  })
  describe('against a real proxied server', function () {
    before(function () {
      var self = this

      this.proxy = null
      this.source = null

      return new Promise(function (resolve) {
        var app = express()

        app.use(function (req, res, next) {
          res.status(200)
          res.set('Content-Type', 'text/plain')
          res.set('X-Source', 'yes!')
          res.end('foo')
        })

        self.source = http.Server(app)

        self.source.listen(0, function () {
          resolve(self.source.address().port)
        })
      }).then(function (sourcePort) {
        return new Promise(function (resolve) {
          var app = express()

          app.use(function (req, res, next) {
            hijackResponse(res, function (err, res) {
              if (err) {
                return next(err)
              }
              res.setHeader('X-Hijacked', 'yes!')
              res.setHeader('transfer-encoding', 'chunked') // not set on > 0.10
              res.removeHeader('Content-Length') // only set on > 0.10
              res.pipe(res)
            })
            next()
          })

          app.use(require('http-proxy-middleware')({
            target: 'http://localhost:' + sourcePort,
            changeOrigin: true
          }))

          self.proxy = http.Server(app)

          self.proxy.listen(0, function () {
            resolve(self.proxy.address().port)
          })
        })
      }).then(function (proxyPort) {
        self.proxyPort = proxyPort
      })
    })
    it('should not mangle response message', function () {
      var self = this
      return new Promise(function (resolve, reject) {
        require('http').get({
          host: 'localhost',
          port: self.proxyPort,
          path: '',
          agent: false
        }, function (res) {
          resolve(res)
        })
      }).then(function (res) {
        return expect(res.headers, 'to exhaustively satisfy', {
          'x-powered-by': 'Express',
          'content-type': /text\/plain/,
          'transfer-encoding': 'chunked',
          date: expect.it('to be a string'),
          connection: 'close',
          'x-source': 'yes!',
          'x-hijacked': 'yes!'
        })
      })
    })
    after(function () {
      this.proxy.close()
      this.source.close()
    })
  })
})


lib/hijackResponse.js
const Transform = require("stream").Transform;
const BrokenPromise = require("./BrokenPromise");

const isHijackedSymbol = Symbol("hijackResponse-isHijacked");

/**
 * Hijack a ServerResponse object, calling you when the headers would otherwise
 * have been sent, and returning you a readable stream, containing what the
 * others thought they wrote to the ServerResponse object, and a writeable
 * stream which will allow you to write to the actual ServerResponse object.
 *
 * @param {import("http").ServerResponse} res A Node.js ServerResponse object
 * @param {Function=} cb A callback - e.g. next when using express.js
 * @return {Promise<{ readable: NodeJS.ReadableStream, writable: NodeJS.WritableStream, destroyAndRestore: Function}>}
 */
module.exports = function hijackResponse(res, cb) {
  const brokenPromise = new BrokenPromise();

  if (res[isHijackedSymbol]) {
    const err = new Error("You cannot hijack a hijacked response.");
    brokenPromise.reject(err);
  }

  // Take a reference to the methods on res that we're going to monkey patch,
  // so we can access them later.
  const __beforePatching = {
    write: res.write,
    end: res.end,
    on: res.on,
    writeHead: res.writeHead
  };

  let implicitHeaderCalled = false;
  function callImplicitHeaderOnce() {
    if (!implicitHeaderCalled) {
      res._implicitHeader();
      implicitHeaderCalled = true;
    }
  }

  const hijackedResponseBody = new Transform({
    transform(chunk, encoding, callback) {
      // Almost a PassThrough Stream. We just need to be able to set a boolean
      // and call a function when the first chunk is being processed.
      callImplicitHeaderOnce();
      callback(null, chunk, encoding);
    },
    flush(cb) {
      callImplicitHeaderOnce();
      cb(null);
    }
  });

  const hijackedResponse = {
    readable: hijackedResponseBody,
    writable: new Proxy(res, {
      get(target, prop) {
        // Make it detectable that we are dealing with a hijacked ServerResponse
        // and not a normal one. Used to prevent attempted hijacks of
        // hijackedResponse.writable.
        if (prop === isHijackedSymbol) {
          return true;
        }

        // Allow access to the original methods from res, before we monkey
        // patched them.
        if (prop in __beforePatching) {
          return __beforePatching[prop].bind(target);
        }

        // Fallback to the value from res.
        return Reflect.get(...arguments);
      }
    }),
    destroyAndRestore() {
      // Restore res, in order to, for example, allow the express errorhandler
      // to use res. (passing errors to next won't work otherwise).
      for (const [methodName, method] of Object.entries(__beforePatching)) {
        res[methodName] = method;
      }
      return hijackedResponseBody.destroy();
    }
  };

  res.on = hijackedResponseBody.on.bind(hijackedResponseBody);
  res.write = hijackedResponseBody.write.bind(hijackedResponseBody);
  res.end = hijackedResponseBody.end.bind(hijackedResponseBody);

  // res.writeHead needs to do a little more than the other monkey-patched
  // methods, which are just forwarding the calls to our stream. writeHead is a
  // method from lib/_http_server.js  in node core.
  //
  // We have to patch the args onto res, as the original writeHead would do,
  // but then not actually write the headers, instead passing control over to
  // the hijacker by resolving the promise that we handed them out.
  res.writeHead = function hijackWriteHead(statusCode, statusMessage, headers) {
    if (typeof headers === "undefined" && typeof statusMessage === "object") {
      headers = statusMessage;
      statusMessage = undefined;
    }
    if (statusCode) {
      res.statusCode = statusCode;
    }
    if (headers) {
      for (const headerName in headers) {
        res.setHeader(headerName, headers[headerName]);
      }
    }
    res.writeHead = __beforePatching.writeHead;

    brokenPromise.resolve(hijackedResponse);
  };

  // Call the callback, if we were passed any. This is likely going to be the
  // next function from an express middleware if provided.
  if (typeof cb === "function") {
    setImmediate(cb);
  }

  return brokenPromise;
};


poc.js
var Writable = require('stream').Writable
var Readable = require('stream').Readable
var Transform = require('stream').Transform

var debug = require('debug')('hijack:debug')
var info = require('debug')('hijack:info')
var log = require('debug')('hijack:log')

var Stream = require('stream').Stream

var convertToUpperCaseTransformStream = new Transform()
convertToUpperCaseTransformStream.count = 0
convertToUpperCaseTransformStream._transform = function (chunk, encoding, cb) {
  if (convertToUpperCaseTransformStream.count === 0) {
    convertToUpperCaseTransformStream.count += 1
    return setTimeout(function () {
      if (encoding === 'buffer') {
        var result = []
        for (var i = 0; i < chunk.length; ++i) {
          var charCode = chunk[i]
          if (charCode > 96 && charCode < 123) {
            charCode = charCode - 32
          }
          result.push(charCode)
        }
        this.push(new Buffer(result))
        return cb()
      }
      return cb(new Error('chunk not a buffer!'))
    }.bind(this), 1000)
  }
  if (encoding === 'buffer') {
    var result = []
    for (var i = 0; i < chunk.length; ++i) {
      var charCode = chunk[i]
      if (charCode > 96 && charCode < 123) {
        charCode = charCode - 32
      }
      result.push(charCode)
    }
    this.push(new Buffer(result))
    return cb()
  }
  return cb(new Error('chunk not a buffer!'))
}

//process.stdin
//  .pipe(convertToUpperCaseTransformStream)
//  .pipe(process.stdout)

function hijack (res, cb) {
  var origWrite = res.write
  var origEnd = res.end
  var origRes = res

  var hijackedRes = new Readable()
  hijackedRes.__proto__ = origRes

  var readableMethods = Object.keys(Readable.prototype)
  readableMethods.forEach(function (method) {
    hijackedRes[method] = Readable.prototype[method].bind(hijackedRes)
  })

  var _readReady = false
  var writeBuffer = []

  function maybeWrite (chunk, writeBuffer) {
    maybeNot = _readReady ? 'not ' : ''
    debug('hijacked data from res. ' + maybeNot + 'buffering')
      if (_readReady) {
        _readReady = hijackedRes.push(chunk)
        return true
      } else {
        writeBuffer.push(chunk)
        return false
      }
  }

  function emptyWriteBuffer () {
    if (_readReady && writeBuffer.length > 0) {
      var newWriteBuffer = []
      writeBuffer.forEach(function (chunk) {
        maybeWrite(chunk, newWriteBuffer)
      })
      writeBuffer = newWriteBuffer
    }
  }

  hijackedRes._read = function (size) {
    debug('_read called %s [from _readReady state %s]', size, _readReady)
    _readReady = true

    emptyWriteBuffer()
    return
  }

  res.write = function (rawChunk, encoding) {
    var chunk = rawChunk
    if (rawChunk !== null && !Buffer.isBuffer(chunk) && encoding !== 'buffer') {
      if (!encoding) {
        chunk = new Buffer(rawChunk)
      } else {
        chunk = new Buffer(rawChunk, encoding)
      }
    }
    debug('res.write called with "%s"', rawChunk ? rawChunk.toString().replace('\n', '\\n') : 'null')
    emptyWriteBuffer()
    return maybeWrite(chunk, writeBuffer)
  }

  res.end = function (chunk, encoding) {
    if (chunk) {
      res.write(chunk, encoding)
    }
    res.write(null)
  }

  hijackedRes.write = function (chunk, encoding) {
    debug('hijackedRes.write called')
    origWrite.call(origRes, chunk, encoding)
  }

  hijackedRes.end = function (chunk, encoding) {
    if (chunk) {
      origWrite.call(origRes, chunk, encoding)
    }
    origEnd.call(origRes)
  }

  return cb(null, hijackedRes)
}

var res = new Writable()

res._write = function (chunk, encoding, next) {
  if (encoding === 'buffer') {
    process.stdout.write(chunk)
    return next()
  }
  throw new Error('not a buffer!')
}



hijack(res, function (err, res) {
  res
    .pipe(convertToUpperCaseTransformStream)
    .pipe(res)
})

hijack(res, function (err, res) {
  res.on('data', function (chunk) {
    console.log('on data', chunk.toString().replace('\n', '\\n'))
    setTimeout(function () { res.write(chunk) }, 2)
  }).on('end', function () {
    console.log('on end')
    setTimeout(function () { res.end() }, 4)
  })
})

// var i = 0
// while ((res.write('foo ') !== false ) || i < 10) {
//   res.write('foo ')
//   i++
// }

res.write('bar')
res.end('\n')


res.on('end', function () {
  console.log('ended')
  process.exit(0)
})
